### PR TITLE
[WIP] Implement spectral Galilean algorithm

### PR DESCRIPTION
**Pull request by oshapoval**
Description edited by @RemiLehe:

# Overview

This pull request implements the Galilean algorithm, as described in [this paper](https://aip.scitation.org/doi/10.1063/1.4964770) and [this paper](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.94.053305).

The aim of the Galilean algorithm is to suppress the Numerical Cherenkov Instability. This is done by integrating the PSATD algorithm is a Galilean frame that moves with the plasma.

Note that the current implementation performs the whole PIC cycle in the Galilean frame (i.e. the particles are pushed with the Galilean velocity for instance). In a future PR, we will shift the boundaries of the domain instead.

# Summary of changes

A new user input variable `v_galilean` is introduced. Whenever one of the components of `v_galilean` is non-zero, then the new `GalileanAlgorithm` (a sub-class of `SpectralAlgorithm`) is used.
Also, the particle pusher has been modified in order to account for the Galilean velocity.
Finally, a new 2D test was added, which checks that the instability is suppressed.

### CODE REVIEW

File: Source/Evolve/WarpXEvolveEM.cpp [hunk 1]
```diff
@@ -156,7 +156,7 @@ WarpX::EvolveEM (int numsteps)
         int num_moved = MoveWindow(move_j);
 
         if (max_level == 0) {
-            int num_redistribute_ghost = num_moved + 1;
+            int num_redistribute_ghost = num_moved + 2;
```
Comment (RemiLehe) [inline Source/Evolve/WarpXEvolveEM.cpp]: Could you replace this by:
```
int num_redistribute_ghost = num_moved + 1;
if ( (v_galilean[0]!=0) or (v_galilean[1]!=0) or (v_galilean[2]!=0) ){
    // In the Galilean algorithm, particles may move by more than one ghost cell
    num_redistribute_ghost += 1;
}
```

File: Docs/source/running_cpp/parameters.rst [hunk 2]
```diff
@@ -672,6 +672,9 @@ Numerics and algorithms
     See `this section of the FFTW documentation <http://www.fftw.org/fftw3_doc/Planner-Flags.html>`__
     for more information.
 
+* ``pstad.v_galilean`` (3 floats, in units of the speed of light)
+    Defines the galilean velocity.
```
Comment (RemiLehe) [inline Docs/source/running_cpp/parameters.rst]: Thanks for updating the documentation. Could you add a bit more info:
- Please mention that the default is 0, and that a non-zero `v_galilean` activates the Galilean algorithm, which suppresses the Numerical Cherenkov Instability in boosted-frame simulation, and add:
```
as explained in :doc:`../theory/boosted_frame`
```
- Please mention that this requires the code to be compiled with `USE_PSATD=TRUE`
- We could also mention that this requires the `direct` deposition (does not work with the Esirkepov algorithm).

File: Examples/Tests/galilean/analysis.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#! /usr/bin/env python
```
Comment (RemiLehe) [inline Examples/Tests/galilean/analysis.py]: Could you add a comment to explain what this test checks?
See for instance: https://github.com/ECP-WarpX/WarpX/blob/master/Examples/Tests/particles_in_PML/analysis.py

**Comment by RemiLehe** [comment]
@oshapoval Thanks a lot for implementing this! I added a few minor comments in the code.

File: Source/WarpX.cpp [hunk 4]
```diff
@@ -199,6 +199,7 @@ WarpX::WarpX ()
 
     costs.resize(nlevs_max);
 
+    //v_galilean.resize(3, 0.0);
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/WarpX.cpp]: Done in dd39e451e85.

File: Source/WarpX.cpp [hunk 5]
```diff
@@ -549,6 +550,9 @@ WarpX::ReadParameters ()
         pp.query("nox", nox_fft);
         pp.query("noy", noy_fft);
         pp.query("noz", noz_fft);
+        pp.queryarr("v_galilean", v_galilean);
+        // Scale the velocity by the speed of light
+        for (int i=0; i<3; i++) v_galilean[i] *= PhysConst::c;
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/WarpX.cpp]: Done in dd39e451e85.

File: Source/WarpX.cpp [hunk 6]
```diff
@@ -678,9 +682,18 @@ WarpX::AllocLevelData (int lev, const BoxArray& ba, const DistributionMapping& d
     // When using subcycling, the particles on the fine level perform two pushes
     // before being redistributed ; therefore, we need one extra guard cell
     // (the particles may move by 2*c*dt)
-    const int ngx_tmp = (maxLevel() > 0 && do_subcycling == 1) ? WarpX::nox+1 : WarpX::nox;
-    const int ngy_tmp = (maxLevel() > 0 && do_subcycling == 1) ? WarpX::noy+1 : WarpX::noy;
-    const int ngz_tmp = (maxLevel() > 0 && do_subcycling == 1) ? WarpX::noz+1 : WarpX::noz;
+    int ngx_tmp = (maxLevel() > 0 && do_subcycling == 1) ? WarpX::nox+1 : WarpX::nox;
+    int ngy_tmp = (maxLevel() > 0 && do_subcycling == 1) ? WarpX::noy+1 : WarpX::noy;
+    int ngz_tmp = (maxLevel() > 0 && do_subcycling == 1) ? WarpX::noz+1 : WarpX::noz;
+
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/WarpX.cpp]: Done in dd39e451e85.

File: Source/Particles/WarpXParticleContainer.cpp [hunk 7]
```diff
@@ -378,6 +378,7 @@ WarpXParticleContainer::DepositCurrent(WarpXParIter& pti,
     // Alternatively, we could define xyzminx from tbx (and the same for 3
     // directions and for jx, jy, jz). This way, sx0 would not be needed.
     // Better for memory? worth trying?
+
     const Dim3 lo = lbound(tilebox);
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/Particles/WarpXParticleContainer.cpp]: Done in dd39e451e85.

File: Source/BoundaryConditions/PML.H [hunk 8]
```diff
@@ -85,6 +85,7 @@ public:
          int ncell, int delta, int ref_ratio,
 #ifdef WARPX_USE_PSATD
          amrex::Real dt, int nox_fft, int noy_fft, int noz_fft, bool do_nodal,
+         //const amrex::Array<amrex::Real,3>& v_galilean, //oshapoval
```
[the remark this answers is not in the document]
Comment (oshapoval) [reply Source/BoundaryConditions/PML.H]: Done in dd39e451e85.

File: Source/Evolve/WarpXEvolveEM.cpp (hunk 1 above)
Comment (oshapoval) [reply Source/Evolve/WarpXEvolveEM.cpp]: Done in dd39e451e85.

File: Docs/source/running_cpp/parameters.rst (hunk 2 above)
Comment (oshapoval) [reply Docs/source/running_cpp/parameters.rst]: Done in dd39e451e85.

File: Examples/Tests/galilean/analysis.py (hunk 3 above)
Comment (oshapoval) [reply Examples/Tests/galilean/analysis.py]: Done in dd39e451e85.

**Comment by RemiLehe** [comment]
This is now superseded by #704

**Pull request closed without merging**